### PR TITLE
Research Models

### PR DESCRIPTION
**Pull request by ujjwal-raizada**

### CODE REVIEW

**Comment by ujjwal-raizada** [comment]
@iamkroot resolve conflict, its your fault.

File: backend/research/models.py [hunk 1]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+    proposer = models.ManyToManyField(Faculty)
+    author = models.ManyToManyField(ResearchScholar)
+    agency = models.CharField(max_length=100)
+    title = models.CharField(max_length=100)
+    pi_copi = models.CharField(max_length=100)
+    date_of_sending = models.DateField()
+    start_date = models.DateField()
+    end_date = models.DateField()
+    status = models.TextField()
+
```
Comment (iamkroot) [inline backend/research/models.py:77]: The relevant fields should be `null=True, blank=True`, as they wouldn't be available at the time of creation. Eg: `dates`

File: backend/research/models.py [hunk 2]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+        (REJECTED, "rejected"),
+        (ACCEPTED, "accepted"),
+    ]
+    author_faculty = models.ManyToManyField(Faculty)
+    author_scholar = models.ManyToManyField(ResearchScholar)
+    title = models.CharField(max_length=100)
+    pub_type = models.CharField(max_length=11, choices=PUB_TYPE_CHOICES)
+    pub_date = models.DateField()
+    status = models.CharField(max_length=20, choices=STATUS_CHOICES)
+    dol_number = models.IntegerField()
```
Comment (iamkroot) [inline backend/research/models.py]:
```suggestion
    doi_number = models.IntegerField()
```

File: backend/research/models.py [hunk 3]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+class Project(models.Model):
+    proposer = models.ManyToManyField(Faculty)
+    author = models.ManyToManyField(ResearchScholar)
+    agency = models.CharField(max_length=100)
+    title = models.CharField(max_length=100)
+    pi_copi = models.CharField(max_length=100)
+    date_of_sending = models.DateField()
+    start_date = models.DateField()
+    end_date = models.DateField()
+    status = models.TextField()
```
Comment (iamkroot) [inline backend/research/models.py]: Needs a `choices` parameter.

File: backend/research/models.py [hunk 4]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+    title = models.CharField(max_length=100)
+    pub_type = models.CharField(max_length=11, choices=PUB_TYPE_CHOICES)
+    pub_date = models.DateField()
+    status = models.CharField(max_length=20, choices=STATUS_CHOICES)
+    dol_number = models.IntegerField()
+    page_number = models.IntegerField()
+
+
+class Project(models.Model):
+    proposer = models.ManyToManyField(Faculty)
```
Comment (iamkroot) [inline backend/research/models.py]:
```suggestion
    proposed_by = models.ManyToManyField(Faculty)
```

File: backend/research/models.py [hunk 5]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+    author = models.ManyToManyField(ResearchScholar)
+    agency = models.CharField(max_length=100)
+    title = models.CharField(max_length=100)
```
Comment (iamkroot) [inline backend/research/models.py]: Increase `max_length`

File: backend/research/models.py (hunk 1 above)
Comment (iamkroot) [inline backend/research/models.py:77]: Add `verbose_name` to the relevant fields. Eg: `pi_copi`, `doi_number`

Comment (iamkroot) [inline backend/research/models.py:77]: Add `Meta` class for the models (See `ResearchScholar` Model)

File: backend/research/models.py [hunk 6]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+
+
+class Publication(models.Model):
+    CONFERENCE = "conference"
+    JOURNAL = "journal"
+    PUB_TYPE_CHOICES = [
+        (CONFERENCE, "conference"),
+        (JOURNAL, "journal"),
+    ]
+    COMMUNICATED = "communicated"
```
Comment (iamkroot) [inline backend/research/models.py]: These strings should be abbreviated. (See `ResearchScholar` model)

File: backend/research/models.py [hunk 7]
```diff
@@ -32,3 +32,41 @@ class ResearchScholar(Student):
...
+    REJECTED = "rejected"
+    ACCEPTED = "accepted"
+    STATUS_CHOICES = [
+        (COMMUNICATED, "communicated"),
+        (REJECTED, "rejected"),
+        (ACCEPTED, "accepted"),
+    ]
+    author_faculty = models.ManyToManyField(Faculty)
+    author_scholar = models.ManyToManyField(ResearchScholar)
```
Comment (iamkroot) [inline backend/research/models.py]: Better to have a ManyToMany with CustomUser

**Review by iamkroot — CHANGES_REQUESTED** [review]

File: backend/research/models.py (hunk 2 above)
Comment (iamkroot) [reply backend/research/models.py]: Short for [`Digital Object Identifier`](https://library.uic.edu/help/article/1966/what-is-a-doi-and-how-do-i-use-them-in-citations)

Comment (iamkroot) [reply backend/research/models.py]: Should be a `CharField`.

**Review by iamkroot — CHANGES_REQUESTED** [review]
Migrations 2 and 3 need to be merged. (Delete both, run `makemigrations` again)
Commits should be squashed. (Only 2 commits needed, `fix typo...`, and `add models...`)

File: backend/research/models.py [hunk 8]
```diff
@@ -32,3 +32,50 @@ class ResearchScholar(Student):
...
+        (COMMUNICATED, "communicated"),
+        (REJECTED, "rejected"),
+        (ACCEPTED, "accepted"),
+    ]
+    author = models.ManyToManyField(CustomUser)
+    title = models.CharField(max_length=100)
+    pub_type = models.CharField(max_length=11, choices=PUB_TYPE_CHOICES)
+    pub_date = models.DateField(null=True, blank=True)
+    status = models.CharField(max_length=20, choices=STATUS_CHOICES)
+    doi_number = models.CharField(max_length=255, null=True, blank=True, verbose_name="Digital Object Indentifier Number")
```
Comment (iamkroot) [inline backend/research/models.py]: too long? i don't think the numbers exceed 20-30 chars. We can keep it at 50 to be safe

File: backend/research/models.py [hunk 9]
```diff
@@ -32,3 +32,50 @@ class ResearchScholar(Student):
...
+        verbose_name = "publication"
+        default_related_name = "publications"
+        ordering = ['-pub_date']
+
+class Project(models.Model):
+    proposed_by = models.ManyToManyField(Faculty)
+    author = models.ManyToManyField(ResearchScholar)
+    agency = models.CharField(max_length=100)
+    title = models.CharField(max_length=300)
+    pi_copi = models.CharField(max_length=100, verbose_name="Principal Investigator/ Co-Principal Investigator")
```
Comment (iamkroot) [inline backend/research/models.py]: line too long. Format file using `black` formatter.

File: backend/research/models.py [hunk 10]
```diff
@@ -32,3 +32,50 @@ class ResearchScholar(Student):
...
+    start_date = models.DateField(null=True, blank=True)
+    end_date = models.DateField(null=True, blank=True)
+    status = models.TextField(null=True, blank=True)
+
+    class Meta:
+        verbose_name = "project"
+        default_related_name = "projects"
+        ordering = ['-date_of_sending', '-start_date']
+
+
```
Comment (iamkroot) [inline backend/research/models.py]: 2 trailing newlines. Should be only 1.

**Review by iamkroot — CHANGES_REQUESTED** [review]

**Review by iamkroot — APPROVED** [review]

**Pull request merged**